### PR TITLE
Non-spatial meshes

### PR DESCRIPTION
**Pull request by franzpoeschel**
Implements https://github.com/openPMD/openPMD-standard/pull/193

TODO:

- [x] Forward/backward compatibility? (of API + of created datasets): No compatibility
- [x] Python docu
- [x] Set openPMD version in written datasets as "2.0"
- [x] Merge #1551 first

### CODE REVIEW

**Comment by franzpoeschel** [comment]
This PR should be split into two parts at some point:

1. Introduction of openPMD 2.0, together with some helper functions
2. Introduction of non-spatial meshes

The first part might then also be relevant for other PRs where breaking changes are introduced (e.g. #1493 use of abbreviated attributes by default in JSON backend)

EDIT: Done #1551

File: src/Mesh.cpp [hunk 1]
```diff
@@ -182,21 +183,161 @@ double Mesh::gridUnitSI() const
...
+         * Since some backends cannot distinguish between vector and
+         * scalar values, the most likely answer here is 1.
+         */
+        return 1;
+    }
+} // namespace
+
+std::vector<double> Mesh::gridUnitSIPerDimension() const
+{
+    return getAttribute("gridUnitSI").get<std::vector<double>>();
```
Comment (franzpoeschel) [inline src/Mesh.cpp]: TODO: introduce the same backwards-compatible logic as for `gridUnitSI` (line 330)

**Comment by ikbuibui** [comment]
I took a look over the PR and it seems okay to me. The interface fits for the binningPlugin use case. However the review isnt very thorough as I am unfamiliar with most of the code base and I don't know enough to follow some workflows.

**Review by ikbuibui — APPROVED** [review]

**Comment by ikbuibui** [comment]
A small comment is that in the binningPlugin in PIConGPU i was using ``std::array<double, 7>`` to hold the dimensionality data. I see that the API requires vectors, even though it assumes the 7 SI Dimensions. Is it really required to be like this?

**Comment by franzpoeschel** [comment]
Thank you for reviewing, Tapish!

> A small comment is that in the binningPlugin in PIConGPU i was using `std::array<double, 7>` to hold the dimensionality data. I see that the API requires vectors, even though it assumes the 7 SI Dimensions. Is it really required to be like this?

Can you point me to where exactly you mean? Within the context of this PR, `std::vector` is used for n-dimensional data, i.e. for a 3-dimensional mesh, you would specify three numbers in `Mesh &setGridUnitSI(std::vector<double> const &gridUnitSI)`.

Beyond that, we have always had two representations for Units: `std::array<double, 7>` and `std::map<UnitDimension, double>`. I have taken the chance of this PR to make this a bit more systematic and provide setters of both kinds, as well as conversion functionality.

With this PR,  `std::vector` may now additionally used to specify the unit per dimension: in `Mesh &setGridUnitDimension(unit_representations::AsArrays const &gridUnitDimension);`, the type `AsArrays` resolves to `std::vector<std::array<double, 7>>`. So, the use of `std::array` mostly stays.

**Comment by ikbuibui** [comment]
Yes, you are right. It was my misunderstanding. Looks good!

**Pull request merged**